core/src/main/java/org/apache/iceberg/encryption/StandardEncryptionManager.java

```diff
@@ -0,0 +1,142 @@
+import org.apache.iceberg.util.ByteBuffers;
+
+public class StandardEncryptionManager implements EncryptionManager {
+  private final transient KeyManagementClient kmsClient;
+  private final String tableKeyId;
+  private final int dataKeyLength;
+
+  private transient volatile SecureRandom lazyRNG = null;
+
+  class StandardEncryptedOutputFile implements EncryptedOutputFile {
+
+    private final OutputFile encryptingOutputFile;
+
+    private final EncryptionKeyMetadata keyMetadata;
+    private final OutputFile rawOutputFile;
+
+    StandardEncryptedOutputFile(
+        OutputFile encryptingOutputFile,
+        EncryptionKeyMetadata keyMetadata,
+        OutputFile rawOutputFile) {
+      this.encryptingOutputFile = encryptingOutputFile;
+      this.keyMetadata = keyMetadata;
+      this.rawOutputFile = rawOutputFile;
+    }
+
+    @Override
+    public OutputFile encryptingOutputFile() {
+      return encryptingOutputFile;
+    }
+
+    @Override
+    public EncryptionKeyMetadata keyMetadata() {
+      return keyMetadata;
+    }
+
+    @Override
+    public OutputFile rawOutputFile() {
+      return rawOutputFile;
+    }
+  }
+
+  /**
+   * @param tableKeyId table encryption key id
+   * @param dataKeyLength length of data encryption key (16/24/32 bytes)
+   * @param kmsClient Client of KMS used to wrap/unwrap keys in envelope encryption
+   */
+  public StandardEncryptionManager(
+      String tableKeyId, int dataKeyLength, KeyManagementClient kmsClient) {
+    Preconditions.checkNotNull(tableKeyId, "Invalid encryption key ID: null");
+    Preconditions.checkNotNull(kmsClient, "Invalid KMS client: null");
+    this.tableKeyId = tableKeyId;
+    this.kmsClient = kmsClient;
+    this.dataKeyLength = dataKeyLength;
+  }
+
+  @Override
+  public EncryptedOutputFile encrypt(OutputFile rawOutput) {
+    ByteBuffer fileDek = ByteBuffer.allocate(dataKeyLength);
+    workerRNG().nextBytes(fileDek.array());
+
+    ByteBuffer aadPrefix = ByteBuffer.allocate(TableProperties.ENCRYPTION_AAD_LENGTH_DEFAULT);
+    workerRNG().nextBytes(aadPrefix.array());
+
+    KeyMetadata encryptionMetadata = new KeyMetadata(fileDek, aadPrefix);
+
+    return new StandardEncryptedOutputFile(
+        new AesGcmOutputFile(rawOutput, fileDek.array(), aadPrefix.array()),
+        encryptionMetadata,
+        rawOutput);
+  }
+
+  @Override
+  public InputFile decrypt(EncryptedInputFile encrypted) {
+    KeyMetadata keyMetadata = KeyMetadata.castOrParse(encrypted.keyMetadata());
+
+    byte[] fileDek = ByteBuffers.toByteArray(keyMetadata.encryptionKey());
+    byte[] aadPrefix = ByteBuffers.toByteArray(keyMetadata.aadPrefix());
+
+    return new AesGcmInputFile(encrypted.encryptedInputFile(), fileDek, aadPrefix);
+  }
+
+  @Override
+  public Iterable<InputFile> decrypt(Iterable<EncryptedInputFile> encrypted) {
+    // Bulk decrypt is only applied to data files. Returning source input files for parquet.
+    return Iterables.transform(encrypted, this::getSourceFile);
```

Comment:
This is incorrect. Based on the comment, I think the intent was to skip creating an `AesGcmInputFile` because you know that this is only called for data files. But that's a bad assumption and there is no guarantee it will be called that way in the future.

Instead, I think this should use a decrypted `InputFile` that can return the underlying `encryptedInputFile` just like the `StandardEncryptedOutputFile` and the key metadata above.